osu.Game.Rulesets.Osu.Tests/TestSceneSliderSnaking.cs

```diff
@@ -0,0 +1,304 @@
+            AddStep("retrieve 3rd slider", () => slider = (DrawableSlider)Player.DrawableRuleset.Playfield.AllHitObjects.Skip(2).First());
+            testDoubleRepeating(true);
+            testDoubleRepeating(false);
+        }
+
+        [TestCase(true)]
+        [TestCase(false)]
+        public void TestArrowStays(bool isHit)
+        {
+            var isSame = isHit ? "is same" : "decreased";
+            var enable = isHit ? "enable" : "disable";
+
+            AddStep($"{enable} autoplay", () => autoplay = isHit);
+            setSnaking(true);
+            base.SetUpSteps();
+
+            addSeekStep(13500);
+            AddStep("retrieve 2nd slider repeat", () =>
+            {
+                var drawable = Player.DrawableRuleset.Playfield.AllHitObjects.Skip(1).First();
+                repeat = drawable.ChildrenOfType<Container<DrawableSliderRepeat>>().First().Children.First();
+            });
+            AddStep("Save repeat vector", () => vector = repeat.Position);
+            addSeekStep(13700);
+            AddAssert($"Repeat vector {isSame}", () => isHit ? Precision.AlmostEquals(vector.X, repeat.X, 1) && Precision.AlmostEquals(vector.Y, repeat.Y, 1) : repeat.X < vector.X && repeat.Y < vector.Y);
+        }
+
+        private void testLinear(bool snaking)
+        {
+            var increased = snaking ? "increased" : "is same";
+
+            setSnaking(snaking);
+            addSeekStep(1800);
+            AddStep("Save end vector", () =>
+            {
+                var body = (PlaySliderBody)slider.Body.Drawable;
+                vector = body.CurrentCurve.Last();
+            });
+            addSeekStep(1900);
+            AddAssert($"End vector {increased}", () =>
+            {
+                var body = (PlaySliderBody)slider.Body.Drawable;
+                var last = body.CurrentCurve.Last();
+                return snaking ? last.X > vector.X && last.Y > vector.Y : last == vector;
+            });
+            addSeekStep(3100);
+            AddStep("Save start vector", () =>
+            {
+                var body = (PlaySliderBody)slider.Body.Drawable;
+                vector = body.CurrentCurve.First();
+            });
+            addSeekStep(3200);
+            AddAssert($"Start vector {increased}", () =>
+            {
+                var body = (PlaySliderBody)slider.Body.Drawable;
+                var first = body.CurrentCurve.First();
+                return snaking ? first.X > vector.X && first.Y > vector.Y : first == vector;
+            });
+        }
+
+        private void testRepeating(bool snaking)
+        {
+            var increased = snaking ? "increased" : "is same";
+            var decreased = snaking ? "decreased" : "is same";
+
+            setSnaking(snaking);
+            addSeekStep(8800);
+            AddStep("Save end vector", () =>
+            {
+                var body = (PlaySliderBody)slider.Body.Drawable;
+                vector = body.CurrentCurve.Last();
```

Comment:
This (and the `First()` pattern) repeats a lot. I'd split a method or helper property here.